enterprise/server/remote_execution/action_merger/action_merger.go

```diff
@@ -201,52 +164,133 @@ func recordSubmitTimeOffsetMetric(hash map[string]string, groupIdForMetrics stri
+		pipe.HSetNX(ctx, forwardKey, firstExecutionSubmitTimeKey, nowString)
+		pipe.HSet(ctx, forwardKey, lastExecutionSubmitTimeKey, nowString)
+		pipe.HIncrBy(ctx, forwardKey, hedgedExecutionCountKey, 0)
+		pipe.HIncrBy(ctx, forwardKey, actionCountKey, 1)
+		pipe.Expire(ctx, forwardKey, queuedExecutionTTL)
+		pipe.Set(ctx, reverseKey, forwardKey, queuedExecutionTTL)
+		_, err = pipe.Exec(ctx)
+		executionID = newExecutionID
+		return err
+	}, forwardKey)
+	if err != nil && err != redis.TxFailedErr {
+		// Unexpected redis error.
+		log.CtxWarningf(ctx, "Unexpected redis error while registering execution for %s: %v", forwardKey, err)
+		return newExecutionID, NEW
 	}
+	if err == redis.TxFailedErr {
+		// The pipeline may have failed because another execution was created
+		// between the initial WATCH and the pipeline execution.
+		executionID, err = rdb.HGet(ctx, forwardKey, executionIDKey).Result()
+		if err != nil {
+			log.CtxWarningf(ctx, "Transaction to register execution for %s failed with %v, but reading the key again failed with %v", forwardKey, redis.TxFailedErr, err)
+			return newExecutionID, NEW
+		}
+	}
+
+	if executionID == "" {
+		log.CtxWarningf(ctx, "Failed to register or read execution ID for %s", forwardKey)
+		return newExecutionID, NEW
+	}
+	if executionID == newExecutionID {
+		// We won the race and created the first execution record.
+		return newExecutionID, NEW
+	}
+
+	// At this point, there is an existing execution in some state. We need to
+	// validate it before deciding whether to merge into it.
 	hash, err := rdb.HGetAll(ctx, forwardKey).Result()
 	if err != nil {
-		log.Debugf("Error reading action-merging state from Redis: %s", err)
-		return "", false, nil
-	}
-	executionID, ok := hash[executionIDKey]
-	if !ok {
-		return "", false, err
+		log.CtxDebugf(ctx, "Error reading action-merging state from Redis: %s", err)
+		return newExecutionID, NEW
 	}
 
 	// Validate that the reverse mapping exists as well. The reverse mapping is
 	// used to delete the pending task information when the task is done.
 	// Bail out if it doesn't exist.
 	err = rdb.Get(ctx, redisKeyForPendingExecutionDigest(executionID)).Err()
 	if err == redis.Nil {
-		return "", false, nil
+		log.CtxWarningf(ctx, "Pending execution %q does not exist in the scheduler", newExecutionID)
+		return newExecutionID, NEW
 	}
 	if err != nil {
-		return "", false, err
+		log.CtxWarningf(ctx, "Unexpected redis error while reading reverse key for existing execution ID %q", newExecutionID)
+		return newExecutionID, NEW
 	}
 
-	// Finally, confirm this execution exists in the scheduler and hasn't been
-	// lost somehow.
-	ok, err = schedulerService.ExistsTask(ctx, executionID)
-	if err != nil {
-		return "", false, err
+	if shouldHedge(hash) {
+		return executionID, HEDGE
+	} else {
+		return executionID, MERGE
 	}
-	if !ok {
-		log.CtxWarningf(ctx, "Pending execution %q does not exist in the scheduler", executionID)
```

Comment:
@tylerwilliams @vadimberezniker I am still seeing this in the logs after https://github.com/buildbuddy-io/buildbuddy/commit/e3c32f85bda07d88a87eb292f3cd2a3f82f85c45 (even though it did get rid of the warning for failed cleanup) and I couldn't figure out why. It seems to be linked to one particular workload.